rsp_jupyter_extensions/handlers/query.py

```diff
@@ -1,112 +1,288 @@
-        derived from the type and value.
+        It will then use the value to resolve the template, and construct
+        a filename resolved under $JUPYTER_SERVER_ROOT (in the RSP, the
+        same as $HOME).  If that file exists, we will return it, on the
+        grounds that the user has done this particular query before and we
+        want to keep any changes made.  Otherwise we will write a file with
+        the query template resolved, so the user can run it to retrieve
+        results.
         """
         input_str = self.request.body.decode("utf-8")
         input_document = json.loads(input_str)
         q_type = input_document["type"]
         q_value = input_document["value"]
-        if q_type != "portal":
-            raise UnsupportedQueryTypeError(f"{q_type} is not a supported query type")
-        q_fn = self._create_portal_query(q_value)
+        q_fn = await self._create_query(q_value, q_type)
         self.write(q_fn)
 
-    def _create_portal_query(self, q_value: str) -> str:
-        # The value should be a URL
-        url = q_value
-        q_id = q_value.split("/")[-1]  # Last component is a unique query ID
-        nb = self._get_portal_query_notebook(url)
-        r_qdir = Path("notebooks") / "queries"
-        qdir = Path(os.getenv("HOME", "")) / r_qdir
-        qdir.mkdir(parents=True, exist_ok=True)
-        fname = f"portal_{q_id}.ipynb"
-        r_fpath = r_qdir / fname
-        fpath = qdir / fname
-        fpath.write_text(nb)
-        retval = {
-            "status": 200,
-            "filename": str(fname),
-            "path": str(r_fpath),
-            "url": (
-                os.environ.get("JUPYTERHUB_SERVICE_PREFIX", "")
-                + "/tree/"
-                + str(r_fpath)
-            ),
-            "body": nb,
-        }
-        return json.dumps(retval)
-
-    def _get_portal_query_notebook(self, url: str) -> str:
+    async def _create_query(self, q_value: str, q_type: str) -> str:
+        match q_type:
+            case "tap":
+                return await self._create_tap_query(q_value)
+            case _:
+                raise UnsupportedQueryTypeError(
+                    f"{q_type} is not a supported query type"
+                )
+
+    async def _create_tap_query(self, q_value: str) -> str:
+        # The value should be a URL or a jobref ID
+        this_rsp = os.getenv("EXTERNAL_INSTANCE_URL", "not-an-rsp")
+        if q_value.startswith(this_rsp):
+            # This looks like a URL
+            url = q_value
+            q_id = q_value.split("/")[-1]  # Last component is the jobref ID
+        else:
+            # It's a raw jobref ID
+            url = f"{this_rsp}/api/tap/async/{q_value}"
+            q_id = q_value
+        fname = (
+            Path(os.getenv("JUPYTER_SERVER_ROOT", ""))
+            / "notebooks"
+            / "queries"
+            / f"tap_{q_id}.ipynb"
+        )
+        if fname.is_file():
+            nb=fname.read_text()
```

Comment:
```suggestion
            nb = fname.read_text()
```

Reply:
I'm wondering how this got by the linter.